### PR TITLE
cas support in edx-platform

### PR DESCRIPTION
**Pull request by Ian2012**
**Description:** Based on the CasBackend of django-cas-ng, we've 
created a backend to support CAS protocol on the edx-platform.

Two custom backends are created, one is just to override the `ticket_validate` 
method on  `CASBackend` and the other is to add support to
CAS into edx-platform as part of the `third_party_auth` app.

**Dependencies:** A PR to upstream creating a new ConfigurationModel
to allow the correct integration with edx-platform

**Installation instructions:** ...

**Testing instructions:**

Work in progress, not testing instructions for now.

**Reviewers:**
- [ ] tag reviewer 
- [ ] tag reviewer 

**Merge checklist:**
- [ ] All reviewers approved
- [ ] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPI after tag-triggered build is 
      finished.
- [ ] Delete working branch (if not needed anymore)

**Author concerns:** List any concerns about this PR - inelegant 
solutions, hacks, quick-and-dirty implementations, concerns about 
migrations, etc.

### CODE REVIEW

File: openedx_cas/apps.py [hunk 1]
```diff
@@ -11,3 +11,12 @@ class OpenedxCasConfig(AppConfig):
...
     name = 'openedx_cas'
+
+    plugin_app = {
+        'settings_config': {
+            'lms.djangoapp': {
+                'test': {'relative_path': 'settings.test'},
+                'common': {'relative_path': 'settings.common'}
+            }
+        },
+    }
```
Comment (mariajgrimaldi) [inline openedx_cas/apps.py]: New line missing. This would be nice in another PR; leave this one just for cas flow changes. Meaning creating a new PR with this change, the change of removing test_models, and setup.py changes. We can merge that PR right away

File: openedx_cas/backends.py [hunk 2]
```diff
@@ -2,9 +2,63 @@
...
+import logging
 
-class CASAuth(BaseAuth):
+UserModel = get_user_model()
+
+logger = logging.getLogger(__name__)
+
+
+class CASBackendOverride(CASBackend):
+
```
Comment (mariajgrimaldi) [inline openedx_cas/backends.py:70]: Missing docstring

File: openedx_cas/backends.py [hunk 3]
```diff
@@ -2,9 +2,63 @@
...
+    def cas_validation(self, request, ticket, service):
```
Comment (mariajgrimaldi) [inline openedx_cas/backends.py:71]: Missing method docstring

File: openedx_cas/backends.py [hunk 4]
```diff
@@ -2,9 +2,63 @@
...
+        client = get_cas_client(service_url=service, request=request)
+        username, attributes, pgtiou = client.verify_ticket(ticket)
+
+        if not username:
+            message = f"Couldn't find username associated with ticket{ticket}"
```
Comment (mariajgrimaldi) [inline openedx_cas/backends.py]: Isn't a data leakage to log the ticket?

File: openedx_cas/backends.py [hunk 5]
```diff
@@ -2,9 +2,63 @@
...
+            logger.error(message)
+            raise AuthFailed(message)
+
+        if attributes and request:
+            request.session['attributes'] = attributes
+
+        if settings.CAS_USERNAME_ATTRIBUTE != 'cas:user' and settings.CAS_VERSION != 'CAS_2_SAML_1_0':
```
Comment (mariajgrimaldi) [inline openedx_cas/backends.py]: I always try to be consistent with my quotes, use double or simple

Comment (mariajgrimaldi) [inline openedx_cas/backends.py]:
```suggestion
        if getattr(settings, "CAS_USERNAME_ATTRIBUTE", "cas:user") != 'cas:user' and settings.CAS_VERSION != 'CAS_2_SAML_1_0':
```
This will avoid errors when the setting is not defined

File: openedx_cas/backends.py [hunk 6]
```diff
@@ -25,7 +79,8 @@ def auth_url(self):
 
         Redirect /CAS-provider-domain/
         """
-        raise NotImplementedError("Not implemented yet")
+        settings.CAS_SERVICE_URL = f'http://{settings.LMS_BASE}/auth/complete/{self.name}/?next=/'
```
Comment (mariajgrimaldi) [inline openedx_cas/backends.py]: Don't try to override settings here. If needed, just suppose that the operator configured the setting correctly. Or define it in common.py

File: openedx_cas/backends.py [hunk 7]
```diff
@@ -36,6 +91,25 @@ def auth_complete(self, *args, **kwargs):
...
-	    https://CAS-provider-domain/auth/complete/?ticket=ticket
+        https://CAS-provider-domain/auth/complete/?ticket=ticket
+        :returns: [User] Authenticated User object or None if authenticate failed
         """
-        raise NotImplementedError("Not implemented yet")
+        request = kwargs['request']
+        ticket = request.GET['ticket']
+        service = settings.CAS_SERVICE_URL
+        response = self.cas_validation(request, ticket, service)
+        logger.info(f"CAS Response: {response}")
```
Comment (mariajgrimaldi) [inline openedx_cas/backends.py]: CAS response for service {LMS_BASE}?

File: openedx_cas/backends.py [hunk 8]
```diff
@@ -36,6 +91,25 @@ def auth_complete(self, *args, **kwargs):
...
+        kwargs.update({'response': response, 'backend': self})
+        return self.strategy.authenticate(*args, **kwargs)
+
+    def get_user_details(self, response):
```
Comment (mariajgrimaldi) [inline openedx_cas/backends.py:129]: Missing docstring

File: openedx_cas/backends.py [hunk 9]
```diff
@@ -36,6 +91,25 @@ def auth_complete(self, *args, **kwargs):
...
+        return {
+            'username': response.get("username", ""),
+            'email': response.get("email", ""),
+            'fullname': response.get("fullname", ""),
+            'first_name': response.get("first_name", ""),
+            'last_name': response.get("last_name", ""),
+        }
```
Comment (mariajgrimaldi) [inline openedx_cas/backends.py]: Try to use double or single quotes (follow in all files)

File: openedx_cas/backends.py [hunk 10]
```diff
@@ -36,6 +91,25 @@ def auth_complete(self, *args, **kwargs):
...
+
+    def get_user_id(self, details, response):
```
Comment (mariajgrimaldi) [inline openedx_cas/backends.py:139]: Missing docstring

File: openedx_cas/settings/common.py [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+Common settings for the Open edX Demo site.
+"""
+
+
+def plugin_settings(settings):
+    """
+    Defines openedx-demo-plugin settings when app is used as a plugin to edx-platform.
+    See: https://github.com/openedx/edx-django-utils/tree/master/edx_django_utils/plugins
+    """
+    settings.CAS_SERVER_URL = "https://ingreso.preprod-ceibal.edu.uy/"
```
Comment (mariajgrimaldi) [inline openedx_cas/settings/common.py]: This shouldn't be here

File: openedx_cas/settings/common.py [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+    settings.CAS_SERVER_LOGIN_URL = "https://ingreso.preprod-ceibal.edu.uy/loginunico/username.xhtml"
```
Comment (mariajgrimaldi) [inline openedx_cas/settings/common.py]: This shouldn't be here. Configure in your local environment

File: openedx_cas/settings/common.py [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+    if "openedx_cas.backends.CASAuth" not in settings.AUTHENTICATION_BACKENDS:
+        settings.AUTHENTICATION_BACKENDS = ["openedx_cas.backends.CASAuth"] + settings.AUTHENTICATION_BACKENDS
+    settings.CAS_CREATE_USER= True
+    settings.CAS_VERSION = 3
```
Comment (mariajgrimaldi) [inline openedx_cas/settings/common.py]: missing end of file line

File: openedx_cas/settings/test.py [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+class SettingsClass:
+    """ dummy settings class """
+
+
+def plugin_settings(settings):
+    """
+    Defines openedx-demo-plugin settings when app is used as a plugin to edx-platform.
+    See: https://github.com/openedx/edx-django-utils/tree/master/edx_django_utils/plugins
+    """
+    settings.OPEN_EDX_VISITOR_ORG = "Public"
```
Comment (mariajgrimaldi) [inline openedx_cas/settings/test.py]: Here you should configure the necessary settings for testing

File: openedx_cas/settings/test.py [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+        'NAME': 'default.db',
+        'USER': '',
+        'PASSWORD': '',
+        'HOST': '',
+        'PORT': '',
+    }
+}
+
+
+SECRET_KEY = 'insecure-secret-key'
```
Comment (mariajgrimaldi) [inline openedx_cas/settings/test.py]: Blank line

File: setup.py [hunk 16]
```diff
@@ -130,9 +130,6 @@ def is_requirement(line):
     entry_points={
         "lms.djangoapp": [
             'openedx_cas = openedx_cas.apps:OpenedxCasConfig',
-        ],
-        "cms.djangoapp": [
-            'openedx_cas = openedx_cas.apps:OpenedxCasConfig',
-        ],
+        ]
```
Comment (mariajgrimaldi) [inline setup.py:133]: Just add files related to the cas flow PR

File: tests/test_backends.py [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+        ...
+
+    def test_get_user_id_doesnt_exists(self):
+        """
+        This method is used to verify the method throws an error when the username is not defined
+
+        Expected behavior:
+        Throws an KeyError exception
+        """
+        ...
```
Comment (mariajgrimaldi) [inline tests/test_backends.py]: Missing tests implementation

**Review by mariajgrimaldi — CHANGES_REQUESTED** [review]

File: openedx_cas/backends.py (hunk 2 above)
Comment (Ian2012) [reply openedx_cas/backends.py:70]: Docstring is there

File: openedx_cas/backends.py (hunk 4 above)
Comment (Ian2012) [reply openedx_cas/backends.py]: @mariajgrimaldi It seems the ticket expiration can be modified, if it's a high value (some minutes) it can dangerous.

Let's remove it for now.

File: openedx_cas/backends.py (hunk 8 above)
Comment (Ian2012) [reply openedx_cas/backends.py:129]: @mariajgrimaldi Corrected

File: openedx_cas/backends.py [hunk 18]
```diff
@@ -1,41 +1,133 @@
...
+from django.contrib.auth import get_user_model
+
+import logging
+
+UserModel = get_user_model()
 
+logger = logging.getLogger(__name__)
 
-class CASAuth(BaseAuth):
+class CASAuth(BaseAuth, CASBackend):
```
Comment (Ian2012) [inline openedx_cas/backends.py:22]: @mariajgrimaldi I've refactored the old implementation (CASOverrideBackend) and mixed into CASAuth

File: openedx_cas/backends.py [hunk 19]
```diff
@@ -1,41 +1,133 @@
...
+        GET https://CAS-service-domain/auth/complete/?ticket=ticket
+        :returns: [User] Authenticated User object or None if authenticate failed
+        """
+
+        request = kwargs["request"]
+        ticket = request.GET["ticket"]
+        response = self.cas_validation(request, ticket, settings.CAS_SERVICE_URL)
+        kwargs.update({"response": response, "backend": self})
+
+        logger.info(f"User {response} logged in with {settings.CAS_SERVER_URL}")
```
Comment (Ian2012) [inline openedx_cas/backends.py:67]: @mariajgrimaldi I'm not sure if this log is useful (was for development purposes). Does the third_party_app or pipeline logs this information?

File: openedx_cas/backends.py [hunk 20]
```diff
@@ -1,41 +1,135 @@
...
+        
+        return self.strategy.authenticate(*args, **kwargs)
+
+    def cas_validation(self, request, ticket, service):
+        """Performs the CAS validation process
```
Comment (mariajgrimaldi) [inline openedx_cas/backends.py]:
```suggestion
        """Performs the CAS ticket validation process.
``

File: openedx_cas/backends.py [hunk 21]
```diff
@@ -1,41 +1,135 @@
...
+        
+        Overrides CASBackend.cas_validation to make it compatible with third_party_auth app
+        
+        Get the CAS_CLIENT based on the settings.CAS_VERSION and verify the ticket
+        which returns the authentication information of the user.
+
+        Parameters:
```
Comment (mariajgrimaldi) [inline openedx_cas/backends.py]:
```suggestion
        Arguments:
```
I try following this convention: https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings or PEP8

File: openedx_cas/backends.py [hunk 22]
```diff
@@ -1,41 +1,135 @@
...
+            request: HttpRequest
+            ticket: str containing the ticket key
+            service: setting used 
+        Returns:
+            is successful:
+                return a dict containing the username
+            otherwhise:
+                raise AuthFailed exception or None when the username attribute is changed
```
Comment (mariajgrimaldi) [inline openedx_cas/backends.py]:
```suggestion
        Returns:
            - A dictionary containing the username when the validation flow succeeds
            - None when the username attribute is changed
            
      Raises:
                AuthFailed exception when...
```
The indentation is all over the place but you get the idea

File: openedx_cas/backends.py (hunk 10 above)
Comment (mariajgrimaldi) [reply openedx_cas/backends.py:139]: Still missing and missing blank line

File: openedx_cas/backends.py [hunk 23]
```diff
@@ -1,41 +1,144 @@
...
+        return {
+            "username": response.get("username", ""),
+            "email": response.get("email", ""),
+            "fullname": response.get("fullname", ""),
+            "first_name": response.get("first_name", ""),
+            "last_name": response.get("last_name", ""),
+        }
+
+    def get_user_id(self, details, response):
+        """We need to override this to allow creating the UserSocialAuth."""
```
Comment (mariajgrimaldi) [inline openedx_cas/backends.py]:
```suggestion
        """Overriding to identify user by username."""
```

File: openedx_cas/backends.py [hunk 24]
```diff
@@ -1,41 +1,144 @@
...
+        return details["username"]
+
+    def auth_html(self):
+        """Override it for future implementations."""
```
Comment (mariajgrimaldi) [inline openedx_cas/backends.py]:
```suggestion
        """Empty override to avoid abstract method quality errors."""
```

**Review by mariajgrimaldi — APPROVED** [review]
